### DISCUSSION TITLE
Drawing Package Options

### DISCUSSION THREAD

**Discussion by lreading** [OWASP/threat-dragon#116] [General] [answered]
[mxgraph](https://github.com/jsGraph/mxgraph) seems as though it may be in flux.  There's a note in the readme that says it's effective EOL, however, there seems to be a community working towards supporting it moving forward.

My question is if we want to invest the time and effort implementing this as the primary drawing package if there's a chance that it's sunsetted and/or not maintained in the near future? Were there any other options being considered, or is this a package that was chosen for specific reasons that we might not be able to find elsewhere?

I'm asking this now as I should be starting the diagram work in the vue-migration branch this week or next. :smiley:

**Comment by jgadsden** [comment, marked as answer]
Hello @lreading - good that you raise this point. The mxgraph community was rocked by this announcement, seems to be that the use of mxgraph in certain domains overlaps with a copyright agreement. This does not affect us, but it means that the general availability of mxgraph as it stood was not sustainable. I am not that clear about what the actual problem was, but it is discussed in :
https://github.com/jsGraph/mxgraph/discussions/23

The open source community have taken on the project at: https://github.com/jsGraph/mxgraph
and I have put forward the need for it to be stable from a Threat Dragon point of view. @coclav kindly made me an admin for mxgraph so that we could help with stability and development

The reason mxgrqph was chosen was because it was fully open source, and provides the drawing package for draw.io or app.diagrams.net . If there is a better drawing package then that would be great ... just that we do not want to go down the JointJS restricted route again :-)

**Reply by lreading** [reply]
Anything with the word "legal" attached scares me a little :sweat_smile: 

I'm not a lawyer, but it looks like the Apache license was modified to prevent it from being used by or integrated with anything owned by Atlassian.  I don't know if the new fork could _legally_ change that part of the license?  I've opened a [discussion](https://github.com/jsGraph/mxgraph/discussions/35) at mxgraph's new home to see if we can get that resolved.  This would prohibit us from adding integrations for Jira, Confluence, or even Bamboo (Atlassian's CI platform, if that's still a thing).

If we continue down the mxgraph (or whatever the name becomes) path, I'd want to make sure we have a way of [addressing the lack of npm package updates](https://github.com/jsGraph/mxgraph/discussions/23#discussioncomment-621353), as you had mentioned.  I've pinged the author of the discussion to see if we can get the re-naming ball rolling. 

Depending on the outcomes of the discussions, we may want to reconsider.  It does look like there's strong community support to keep mxgraph alive, which is awesome! I am, however, hesitant to add an unmaintained package to the 2.0 migration.  There's other work I can do before this becomes a blocker, I think it's worth seeing where the discussion goes. :smiley:

**Reply by jgadsden** [reply]
Yes, agreed, if you can hold off integrating mxgraph for as long as you can then that seems sensible.

I could create a new mxgraph package right away, I have admin rights to the repo and have done this sort of thing before. I guess I was waiting to see if we/they fixed on a name

**Reply by lreading** [reply]
No need to rush anything.  Per our side conversation, I'll be exploring what other open source options are out there.  It does seem as though some maintainers are aware of the license issue and have been putting thought into how to move it forward.  Hopefully the name gets sorted out soon enough.  Those would be the two biggest blockers though. Let's give it a bit of time and see where it goes. :smile:

**Reply by lreading** [reply]
I took a look around and there really are not many great options out there.  I did, however, find one option that looks quite promising: https://github.com/jerosoler/Drawflow

This library already has a vue component, is licensed under MIT and seems to be actively maintained.  A demo is here: https://jerosoler.github.io/Drawflow/

I'll need to play with the API a bit to make sure it'll suit Threat Dragon's needs.  The fact that it's already wrapped in a Vue component is a huge selling point for me at least, as it's less work on implementing the graphing library.

**Reply by jgadsden** [reply]
Yes, DrawFlow looks interesting. The demo is fairly restricted, I was not able to select multiple components, and maybe it could be extended for our purposes.

**Reply by lreading** [reply]
I think it's worth trying out.  I'll try to get together a proof of concept on a new branch so we can see if it does everything we need.  I'll need to lean heavily on your experience in determining if it's a good fit. :smile:

**Reply by lreading** [reply]
I looked at drawflow a bit more and it is missing a few features that we would need (including directional arrows). 
I also found https://github.com/antvis/x6 which looks like a fantastic option based on the demo, with one caveat: The documentation is in a language I cannot read/understand.  That would make it a serious challenge to implement.... ~I am thinking that we should continue down the path of mxgraph and continue our efforts to ensure that momentum carries on over there.~

~Taking a look at the dependencies of mxgraph, there really aren't any external dependencies.  There are also no known vulnerabilities against it at the moment.  I think to get started on diagramming, I will just use the existing mxgraph and start wrapping it in vue components.  While I'd like to wrap the entire package in vue components, I think that's too ambitious of a goal.  I think the best approach is going to be creating components within td.vue that will satisfy the needs of Threat Dragon.~

~I'm hoping the API for it does not change with version 5 (maybe 5? pending [discussion](https://github.com/jsGraph/mxgraph/discussions/41), though it has been noted that there will be [breaking changes](https://github.com/jsGraph/mxgraph/discussions/18#discussioncomment-620519).~

As it would turn out, mxGraph, in its current state, does not play nicely with webpack and is difficult to develop against without the proper typings.  With that said, X6's documentation seems very thorough, and it has a native vue component which has the desired typings.  I'm thinking I should power through with the help of google translate and get a PoC done with X6.

**Reply by jgadsden** [reply]
Antvis is looking good, and I will go with whatever you think best on the drawing package. My efforts in getting mxGraph/maxGraph stable are entirely self-interested ... if we were going to use it for Threat Dragon then I wanted it maintained.

Antvis on the other hand is in active development with almost continuous release, so looks like a much better solution - thanks @lreading for finding this
